chain/chain/src/runtime/mod.rs

```diff
@@ -777,23 +777,38 @@ impl RuntimeAdapter for NightshadeRuntime {
             usize::MAX
         };
 
-        // In general, we limit the number of transactions via send_fees.
-        // However, as a second line of defense, we want to limit the byte size
-        // of transaction as well. Rather than introducing a separate config for
-        // the limit, we compute it heuristically from the gas limit and the
-        // cost of roundtripping a byte of data through disk. For today's value
-        // of parameters, this corresponds to about 13megs worth of
-        // transactions.
-        let size_limit = transactions_gas_limit
-            / (runtime_config.wasm_config.ext_costs.gas_cost(ExtCosts::storage_write_value_byte)
-                + runtime_config.wasm_config.ext_costs.gas_cost(ExtCosts::storage_read_value_byte));
+        let size_limit: u64 =
+            if checked_feature!("stable", WitnessTransactionLimits, protocol_version) {
+                // Sum of transactions in the previous and current chunks should not exceed the limit.
+                // Witness keeps transactions from both previous and current chunk, so we have to limit the sum of both.
+                runtime_config
+                    .max_transactions_size_in_witness
+                    .saturating_sub(chunk.prev_chunk_transactions_size) as u64
+            } else {
+                // In general, we limit the number of transactions via send_fees.
+                // However, as a second line of defense, we want to limit the byte size
+                // of transaction as well. Rather than introducing a separate config for
+                // the limit, we compute it heuristically from the gas limit and the
+                // cost of roundtripping a byte of data through disk. For today's value
+                // of parameters, this corresponds to about 13megs worth of
+                // transactions.
+                transactions_gas_limit
+                    / (runtime_config
+                        .wasm_config
+                        .ext_costs
+                        .gas_cost(ExtCosts::storage_write_value_byte)
+                        + runtime_config
+                            .wasm_config
+                            .ext_costs
+                            .gas_cost(ExtCosts::storage_read_value_byte))
+            };
         // for metrics only
         let mut rejected_due_to_congestion = 0;
         let mut rejected_invalid_tx = 0;
         let mut rejected_invalid_for_chain = 0;
 
         // Add new transactions to the result until some limit is hit or the transactions run out.
-        loop {
+        'add_txs_loop: while let Some(transaction_group_iter) = transaction_groups.next() {
```

Comment:
To be honest I'm not a fan of the named loop syntax, how do you feel about it? I think an alternative would be to use a loop break variable for this. It's just personal preference so feel free to ignore, I'm mostly curious about your thoughts having worked with that code.

Reply:
Hmm I really like it, IMO it's one of the nicer features of Rust.

I like that it explicitly says "I want to break out of this loop". Using a variable introduces additional logic that has to be tracked, I have to make sure that this variable correctly interacts with all the breaks, continues, etc. Named loops avoid all of this headache.

Reply:
Sounds good, as long as it works for you I'm fine with that. I'll just need to get used to it :)